software/base/src/test/java/brooklyn/entity/basic/SoftwareProcessEntityTest.java

```diff
@@ -267,6 +275,87 @@ public void testBasicSoftwareProcessStopsProcess() throws Exception {
     }
     
     @Test
+    public void testBasicSoftwareProcessCanStop() {
+        Object[] matrix = new Object[] {
+                StopMode.ALWAYS, true, true,
+                StopMode.ALWAYS, false, true,
+                StopMode.IF_NOT_STOPPED, true, false,
+                StopMode.IF_NOT_STOPPED, false, true,
+                StopMode.NEVER, true, false,
+                StopMode.NEVER, false, false
+        };
+        for (int i = 0; i < matrix.length; i+=3) {
+            StopMode mode = (StopMode)matrix[i];
+            boolean isEntityStopped = (Boolean)matrix[i+1];
+            boolean expected = (Boolean)matrix[i+2];
+            boolean canStop = MachineLifecycleEffectorTasks.canStop(mode, isEntityStopped);
+            assertEquals(canStop, expected);
```

Comment:
Thanks for the data providers pointer, changed the code to use them. Logs the failing mode parameters automatically.